### DISCUSSION TITLE
release_4.3 Linux渲染崩溃，tgfx bug定位及解决。

### DISCUSSION THREAD

**Discussion by qingpingwang** [Tencent/libpag#2259] [General] [answered]
### 问题描述
Linux服务端渲染部分pag会导致崩溃，必现。经定位是tgfx的内存读写越界问题。
[问题描述](https://github.com/Tencent/libpag/discussions/2254)
### 原因定位
RasterPixelBuffer创建的大小是画布大小，当文字存在缩放动画时，数组访问并没有处理越界问题。
导致FTMask.cpp中SpanFunc函数生成文本的时候可能访问越界（超出画布的文字会访问读写上一行或者下一行，效果也是不正确的。超出画布申请的buffer则崩溃。）
### 代码修复

/////////////////////////////////////////////////////////////////////////////////////////////////
//
//  Tencent is pleased to support the open source community by making tgfx available.
//
//  Copyright (C) 2023 THL A29 Limited, a Tencent company. All rights reserved.
//
//  Licensed under the BSD 3-Clause License (the "License"); you may not use this file except
//  in compliance with the License. You may obtain a copy of the License at
//
//      https://opensource.org/licenses/BSD-3-Clause
//
//  unless required by applicable law or agreed to in writing, software distributed under the
//  license is distributed on an "as is" basis, without warranties or conditions of any kind,
//  either express or implied. see the license for the specific language governing permissions
//  and limitations under the license.
//
/////////////////////////////////////////////////////////////////////////////////////////////////

#include "FTMask.h"
#include "FTPath.h"
#include "tgfx/core/Pixmap.h"

namespace tgfx {
static const FTLibrary& GetLibrary() {
  static const auto& library = *new FTLibrary;
  return library;
}

static void Iterator(PathVerb verb, const Point points[4], void* info) {
  auto path = reinterpret_cast<FTPath*>(info);
  switch (verb) {
    case PathVerb::Move:
      path->moveTo(points[0]);
      break;
    case PathVerb::Line:
      path->lineTo(points[1]);
      break;
    case PathVerb::Quad:
      path->quadTo(points[1], points[2]);
      break;
    case PathVerb::Cubic:
      path->cubicTo(points[1], points[2], points[3]);
      break;
    case PathVerb::Close:
      path->close();
      break;
  }
}

std::shared_ptr<Mask> Mask::Make(int width, int height, bool tryHardware) {
  auto pixelRef = PixelRef::Make(width, height, true, tryHardware);
  if (pixelRef == nullptr) {
    return nullptr;
  }
  pixelRef->clear();
  return std::make_shared<FTMask>(std::move(pixelRef));
}

struct RasterTarget {
  unsigned char* origin;
  int pitch;
  int rows;
  const uint8_t* gammaTable;
};

static void SpanFunc(int y, int, const FT_Span* spans, void* user) {
  auto* target = reinterpret_cast<RasterTarget*>(user);
  auto offsetStart = std::max(int(spans->x), 0);
  // 越界
  if (offsetStart >= target->pitch || y >= target->rows || y < 0) {
    return;
  }
  auto* q = target->origin - target->pitch * y + offsetStart;
  auto c = spans->coverage;
  if (target->gammaTable) {
    c = target->gammaTable[c];
  }

  auto offsetEnd = std::min(std::max(spans->x + spans->len, offsetStart), target->pitch);
  auto aCount = offsetEnd - offsetStart;
  /**
   * For small-spans it is faster to do it by ourselves than calling memset.
   * This is mainly due to the cost of the function call.
   */
  switch (aCount) {
    case 7:
      *q++ = c;
    case 6:
      *q++ = c;
    case 5:
      *q++ = c;
    case 4:
      *q++ = c;
    case 3:
      *q++ = c;
    case 2:
      *q++ = c;
    case 1:
      *q = c;
    case 0:
      break;
    default:
      memset(q, c, aCount);
  }
}

void FTMask::onFillPath(const Path& path, const Matrix& matrix, bool needsGammaCorrection) {
  if (path.isEmpty()) {
    return;
  }
  auto pixels = pixelRef->lockWritablePixels();
  if (pixels == nullptr) {
    return;
  }
  const auto& info = pixelRef->info();
  auto finalPath = path;
  auto totalMatrix = matrix;
  totalMatrix.postScale(1, -1);
  totalMatrix.postTranslate(0, static_cast<float>(pixelRef->height()));
  finalPath.transform(totalMatrix);
  auto bounds = finalPath.getBounds();
  bounds.roundOut();
  markContentDirty(bounds, true);
  FTPath ftPath = {};
  finalPath.decompose(Iterator, &ftPath);
  ftPath.setFillType(path.getFillType());
  auto outlines = ftPath.getOutlines();
  auto ftLibrary = GetLibrary().library();
  auto buffer = static_cast<unsigned char*>(pixels);
  int rows = info.height();
  int pitch = static_cast<int>(info.rowBytes());
  RasterTarget target{};
  target.origin = buffer + (rows - 1) * pitch;
  target.pitch = pitch;
  target.rows = rows;
  if (needsGammaCorrection) {
    target.gammaTable = PixelRefMask::GammaTable().data();
  } else {
    target.gammaTable = nullptr;
  }
  FT_Raster_Params params;
  params.flags = FT_RASTER_FLAG_AA | FT_RASTER_FLAG_DIRECT;
  params.gray_spans = SpanFunc;
  params.user = &target;
  for (auto& outline : outlines) {
    FT_Outline_Render(ftLibrary, &(outline->outline), &params);
  }
  pixelRef->unlockPixels();
}
}  // namespace tgfx

**Comment by kevingpqi123** [comment, marked as answer]
这个问题我们上个月已经修复了：https://github.com/Tencent/tgfx/commit/a0115cde400a51b8b9a0c9206dec3391dd461e15
你可以升级到 release/4.3 的最新代码
